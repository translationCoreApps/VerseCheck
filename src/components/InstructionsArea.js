import React from 'react';
import PropTypes from 'prop-types';
import style from '../css/Style';
export const QuoatationMarks = ({ children }) => <strong style={{ color: 'var(--accent-color)' }}>{'"'}{children}{'"'}</strong>;
export const Ellipsis = () => <strong style={{ color: 'var(--accent-color)' }}>{` ${ELLIPSIS} `}</strong>;

let InstructionsArea = ({
  alignedGLText,
  selectionsReducer,
  dontShowTranslation,
  verseText,
  mode,
  translate
}) => {

  if (!verseText) {
    return (
      <div style={style.InstructionsArea}>
        <span>{translate("empty_verse")}</span><br />
      </div>
    );
  }

  if (selectionsReducer.selections.length === 0 && dontShowTranslation) {
    return (
      <div style={style.InstructionsArea}>
        <span>{translate("no_selection")}</span><br />
      </div>
    );
  }

  if (mode === 'select') {
    return (
      <div style={style.InstructionsArea}>
        <span>{translate("please_select")}</span><br />
        <span>
          <strong style={{ color: 'var(--accent-color)' }}>
            {`"${alignedGLText}"`}
          </strong>
        </span><br />
      </div>
    );
  }

  return (
    <div style={style.InstructionsArea}>
      <span>
        <strong style={{ color: 'var(--accent-color)' }}>
          {`"${alignedGLText}"`}
        </strong>
      </span><br />
      <span>{translate("translated_as")}</span><br />
      <span>{renderTextSelection(selectionsReducer.selections, verseText)}</span>
    </div>
  );
};

InstructionsArea.propTypes = {
  translate: PropTypes.func.isRequired,
  alignedGLText: PropTypes.string.isRequired,
  selectionsReducer: PropTypes.object.isRequired,
  dontShowTranslation: PropTypes.bool,
  verseText: PropTypes.string.isRequired,
  mode: PropTypes.string
};

export default InstructionsArea;

const ELLIPSIS = 'â€¦';
/**
 * This is a helper method to determine if the selection needs an ellipsis in
 * between the selected words or not.
 * @param {Array} selections - Array of word objects that the user selected
 * @param {string} verseText - The entire verse string from the current check
 * @returns {boolean} - Whether or not the View should display an ellipsis
 */
function shouldRenderEllipsis(selections, verseText) {
  /** Need to get the the words and occurrence of the selected edge words */
  const endSelectedWord = selections[selections.length - 1].text.trim();
  const endSelectedWordOccurrence = selections[selections.length - 1].occurrence;
  const beginningSelectedWord = selections[0].text.trim();
  const beginningSelectedWordOccurrence = selections[0].occurrence;

  /** Using the occurrences to get the actual index of the word vs
   *  the first time it appears in verse text */
  const indexOfBeginningSelection = verseText.split(beginningSelectedWord, beginningSelectedWordOccurrence).join(beginningSelectedWord).length;
  const indexOfEndSelection = verseText.split(endSelectedWord, endSelectedWordOccurrence).join(endSelectedWord).length;

  /** Checking the text in between selected words for a non space character */
  const textBetweenSelection = verseText.substring(indexOfBeginningSelection + beginningSelectedWord.length, indexOfEndSelection);
  /** If the end index is the same as the beginning then it is the first word */
  return (indexOfEndSelection !== indexOfBeginningSelection) && textBetweenSelection.match(/\S/);
}

function renderTextSelection(selections, verseText) {
  if (shouldRenderEllipsis(selections, verseText)) {
    return (
      <QuoatationMarks>
        {selections[0].text.trim()}
        <Ellipsis />
        {selections[selections.length - 1].text.trim()}
      </QuoatationMarks>
    );
  } else {
    return (
      <QuoatationMarks>
        {selections.map((selection, index) =>
          <span key={index}>
            <strong style={{ color: 'var(--accent-color)' }}>
              {`${selection.text.trim()}`}
            </strong>
            {selections[index + 1] ? <span>{" "}</span> : null}
          </span>
        )}
      </QuoatationMarks>
    );
  }
}
