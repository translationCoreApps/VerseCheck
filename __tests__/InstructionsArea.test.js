/* eslint-env jest */
import React from 'react';
import InstructionsArea, { QuoatationMarks } from '../src/components/InstructionsArea';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('InstructionsArea component Tests', () => {
  let props;
  const mock_translate = (text) => { return text; };

  beforeEach(()=> {
    props = {
      translate: mock_translate,
      alignedGLText: 'God\'s',
      selectionsReducer: {
        selections: [{text:'selected text'}]
      },
      dontShowTranslation: false,
      verseText: 'This is the verse text',
      mode: 'select'
    };
  });

  test('Check InstructionsArea component mode select', () => {
    const wrapper = shallow(<InstructionsArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual('please_select"'+props.alignedGLText+'"');
  });

  test('Test InstructionsArea component default mode)', () => {
    props.mode = 'default';
    const wrapper = mount(<InstructionsArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual('"'+props.alignedGLText+'"translated_as"'+props.selectionsReducer.selections[0].text+'"');
  });

  test('Test InstructionsArea component default mode with 1 selection)', () => {
    props.mode = 'default';
    props.selectionsReducer.selections = [{text:'selected text'}, {text:'another selected text'}];
    const wrapper = mount(<InstructionsArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual(`"${props.alignedGLText}"translated_as"selected text another selected text"`);
  });

  test('Test InstructionsArea component default mode with multiple selections)', () => {
    props.mode = 'default';
    props.selectionsReducer.selections = [{text:'selected text'}, {text:'another selected text'}, {text:'Alot of selections'}];
    const wrapper = mount(<InstructionsArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual(`"${props.alignedGLText}"translated_as"selected text â€¦ Alot of selections"`);
  });


  test('Test InstructionsArea component no selections)', () => {
    props.selectionsReducer.selections = [];
    props.dontShowTranslation = true;
    const wrapper = shallow(<InstructionsArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual('no_selection');
  });

  test('Test InstructionsArea component no verse text)', () => {
    props.verseText = '';
    const wrapper = shallow(<InstructionsArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual('empty_verse');
  });
});
