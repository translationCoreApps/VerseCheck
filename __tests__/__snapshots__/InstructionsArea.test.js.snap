// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructionsArea component Tests Check InstructionsArea component mode select 1`] = `
<div
  style={
    Object {
      "overflow": "auto",
      "padding": "5px",
      "textAlign": "center",
    }
  }
>
  <span>
    please_select
  </span>
  <br />
  <span>
    <strong
      style={
        Object {
          "color": "var(--accent-color)",
        }
      }
    >
      "God's"
    </strong>
  </span>
  <br />
</div>
`;

exports[`InstructionsArea component Tests Test InstructionsArea component default mode with 1 selection) 1`] = `
<InstructionsArea
  alignedGLText="God's"
  dontShowTranslation={false}
  mode="default"
  selectionsReducer={
    Object {
      "selections": Array [
        Object {
          "occurrence": 1,
          "occurrences": 1,
          "text": "Mungu",
        },
        Object {
          "occurrence": 1,
          "occurrences": 2,
          "text": "na",
        },
      ],
    }
  }
  translate={[Function]}
  verseText="Lakini huruma ya Mungu na wetu na upendo wake kwa wanadamu ulipoonekana, "
>
  <div
    style={
      Object {
        "overflow": "auto",
        "padding": "5px",
        "textAlign": "center",
      }
    }
  >
    <span>
      <strong
        style={
          Object {
            "color": "var(--accent-color)",
          }
        }
      >
        "God's"
      </strong>
    </span>
    <br />
    <span>
      translated_as
    </span>
    <br />
    <span>
      <QuoatationMarks>
        <strong
          style={
            Object {
              "color": "var(--accent-color)",
            }
          }
        >
          "
          <span
            key="0"
          >
            <strong
              style={
                Object {
                  "color": "var(--accent-color)",
                }
              }
            >
              Mungu
            </strong>
            <span>
               
            </span>
          </span>
          <span
            key="1"
          >
            <strong
              style={
                Object {
                  "color": "var(--accent-color)",
                }
              }
            >
              na
            </strong>
          </span>
          "
        </strong>
      </QuoatationMarks>
    </span>
  </div>
</InstructionsArea>
`;

exports[`InstructionsArea component Tests Test InstructionsArea component default mode with multiple selections) 1`] = `
<InstructionsArea
  alignedGLText="God's"
  dontShowTranslation={false}
  mode="default"
  selectionsReducer={
    Object {
      "selections": Array [
        Object {
          "occurrence": 1,
          "occurrences": 1,
          "text": "Mungu",
        },
        Object {
          "occurrence": 2,
          "occurrences": 2,
          "text": "na",
        },
      ],
    }
  }
  translate={[Function]}
  verseText="Lakini huruma ya Mungu na wetu na upendo wake kwa wanadamu ulipoonekana, "
>
  <div
    style={
      Object {
        "overflow": "auto",
        "padding": "5px",
        "textAlign": "center",
      }
    }
  >
    <span>
      <strong
        style={
          Object {
            "color": "var(--accent-color)",
          }
        }
      >
        "God's"
      </strong>
    </span>
    <br />
    <span>
      translated_as
    </span>
    <br />
    <span>
      <QuoatationMarks>
        <strong
          style={
            Object {
              "color": "var(--accent-color)",
            }
          }
        >
          "
          Mungu
          <Ellipsis>
            <strong
              style={
                Object {
                  "color": "var(--accent-color)",
                }
              }
            >
               â€¦ 
            </strong>
          </Ellipsis>
          na
          "
        </strong>
      </QuoatationMarks>
    </span>
  </div>
</InstructionsArea>
`;

exports[`InstructionsArea component Tests Test InstructionsArea component default mode) 1`] = `
<InstructionsArea
  alignedGLText="God's"
  dontShowTranslation={false}
  mode="default"
  selectionsReducer={
    Object {
      "selections": Array [
        Object {
          "occurrence": 1,
          "occurrences": 1,
          "text": "Mungu",
        },
      ],
    }
  }
  translate={[Function]}
  verseText="Lakini huruma ya Mungu na wetu na upendo wake kwa wanadamu ulipoonekana, "
>
  <div
    style={
      Object {
        "overflow": "auto",
        "padding": "5px",
        "textAlign": "center",
      }
    }
  >
    <span>
      <strong
        style={
          Object {
            "color": "var(--accent-color)",
          }
        }
      >
        "God's"
      </strong>
    </span>
    <br />
    <span>
      translated_as
    </span>
    <br />
    <span>
      <QuoatationMarks>
        <strong
          style={
            Object {
              "color": "var(--accent-color)",
            }
          }
        >
          "
          <span
            key="0"
          >
            <strong
              style={
                Object {
                  "color": "var(--accent-color)",
                }
              }
            >
              Mungu
            </strong>
          </span>
          "
        </strong>
      </QuoatationMarks>
    </span>
  </div>
</InstructionsArea>
`;

exports[`InstructionsArea component Tests Test InstructionsArea component no selections) 1`] = `
<div
  style={
    Object {
      "overflow": "auto",
      "padding": "5px",
      "textAlign": "center",
    }
  }
>
  <span>
    no_selection
  </span>
  <br />
</div>
`;

exports[`InstructionsArea component Tests Test InstructionsArea component no verse text) 1`] = `
<div
  style={
    Object {
      "overflow": "auto",
      "padding": "5px",
      "textAlign": "center",
    }
  }
>
  <span>
    empty_verse
  </span>
  <br />
</div>
`;
